pydicom/fileset.py

```diff
@@ -0,0 +1,2861 @@
+        The current node's child nodes (if any)
+    instance : FileInstance or None
+        If the current node is a leaf node, a
+        :class:`~pydicom.fileset.FileInstance` for the corresponding SOP
+        Instance.
+    """
+    def __init__(self, record: Optional[Dataset] = None) -> None:
+        """Create a new ``RecordNode``.
+
+        Parameters
+        ----------
+        record : pydicom.dataset.Dataset, optional
+            A *Directory Record Sequence's* directory record.
+        """
+        self.children = []
+        self.instance = None
+        self._parent = None
+        self._record = None
+
+        if record:
+            self._set_record(record)
+
+        # When the record is encoded as part of the *Directory Record Sequence*
+        #   this is the offset to the start of the sequence item containing
+        #   the record - not guaranteed to be up-to-date
+        self._offset = 0
+        # The offset to the start of the encoded record's *Offset of the
+        #   Next Directory Record* and *Offset of Referenced Lower Level
+        #   Directory Entity* values - use _encode_record() to set them
+        self._offset_next = 0
+        self._offset_lower = 0
+
+    def add(self, leaf: "RecordNode") -> None:
+        """Add a leaf to the tree.
+
+        Parameters
+        ----------
+        leaf : pydicom.fileset.RecordNode
+            A leaf node (i.e. one with a
+            :class:`~pydicom.fileset.FileInstance`) to be added to the tree
+            (if not already present).
+        """
+        # Move up to the branch's furthest ancestor with a directory record
+        node = leaf.root
+        if node is self:
+            node = node.children[0]
+
+        # Move back down, inserting at the point where the node is unique
+        current = self.root
+        while node in current and node.children:
+            current = current[node]
+            node = node.children[0]
+
+        node.parent = current
+
+    @property
+    def ancestors(self) -> List["RecordNode"]:
+        """Return a list of the current node's ancestors, ordered from nearest
+        to furthest.
+        """
+        return [nn for nn in self.reverse() if nn is not self]
+
+    @property
+    def component(self) -> str:
+        """Return a File ID component as :class:`str` for the current node."""
+        if self.is_root:
+            raise ValueError(
+                "The root node doesn't contribute a File ID component"
+            )
+
+        prefix = _PREFIXES[self.record_type]
+        if self.record_type == "PRIVATE":
+            prefix = f"{prefix}{self.depth}"
+
+        chars = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+        if not self.file_set._use_alphanumeric:
+            chars = chars[:10]
+
+        suffix = ""
+        n = self.index
+        b = len(chars)
+        while n:
+            suffix += chars[n % b]
+            n //= b
+
+        idx = f"{suffix[::-1]:>0{8 - len(prefix)}}"
+
```

Comment:
I think it may be possible and it's what I started out thinking I was going to do. But either I'd just end up creating a new generator for every node, or I'd have to have a bunch of them and track their use at each level of the hierarchy and that was just going to be far too complicated. 

I'm quite happy with how this solution just fell out once I had the node depth and sibling index. It automatically adjusts the File IDs as the tree changes and its conceptually simple, whereas if I'd used the generator I was probably going to have to use `send()` or restart them after every change.